README.md
# Borachio

Native Scala mocking.

## Borachio is now [ScalaMock](http://scalamock.org/)

This repository is now mothballed. Future development will happen [here](/paulbutcher/scalamock).

RELEASE_NOTES.md
Version 1.5:

- Fix mocks that return mocks

Version 1.4:

- Add `stubs` method as syntactic sugar for `expects ... anyNumberOfTimes`
- Add `inAnyOrder` and allow ordered and unordered expectations to be arbitrarily nested.
- Add `where` for predicate matching and `onCall` for computed return values.

Version 1.3:

- Switch to using thread context class loader when creating proxies by default
- Add ability to override proxy class loader strategy
- Add support for Scala 2.9.1
- Switch to using sbt 0.10
- Split test framework support out into separate libraries

Version 1.2:

- Add support for Scala 2.9.0-1
- Give a slightly nicer error if withExpectations isn't used

Version 1.1:

- Add the CallLogging trait

Version 1.0:

- Scala 2.9.0 support
- Test frameworks are now optional dependencies

Version 0.8:

- Integration with specs2 - [Daniel Westheide](https://github.com/dwestheide)

Version 0.7:

- Stop expectation errors being lost when the SUT swallows exceptions

Version 0.6:

- Add VerboseErrors trait
- Add withExpectations for JUnit3
- Fix anyNumberOfTimes, which didn't work because the range (0 to scala.Int.MaxValue) is actually empty

Version 0.5:

- Improved error messages for inSequence expectations

Version 0.4:

- Add Junit3 support

Version 0.3:

- Rethink argument matching to do away with the need for the ^ method
- Improve error messages

Version 0.2:

- Add support for wildcard and epsilon argument matchers

Version 0.1:

- Initial release


core_tests/src/test/scala/ProxyTest.scala
// Copyright (c) 2011 Paul Butcher
// 
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
// 
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
// 
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.

package com.borachio

import org.scalatest.WordSpec

trait Trait1 { def method1(x: Int, y: String): String }
trait Trait2
trait Trait3

abstract class Class1 extends Trait1 with Trait2

class ProxyTest extends WordSpec {
  
  def unimplemented(proxy: AnyRef, name: Symbol, args: Array[AnyRef]): AnyRef =
    throw new RuntimeException("unimplemented")
  
  "A proxy" should {
    
    "implement a single interface" in {
      val p = Proxy.create(threadContextClassLoaderStrategy, classOf[Trait1]) {
        unimplemented _
      }
      assert(p.isInstanceOf[Trait1])
    }
    
    "implement multiple interfaces" in {
      val p = Proxy.create(threadContextClassLoaderStrategy, classOf[Trait1], classOf[Trait2], classOf[Trait3]) {
        unimplemented _
      }
      assert(p.isInstanceOf[Trait1])
      assert(p.isInstanceOf[Trait2])
      assert(p.isInstanceOf[Trait3])
    }
    
    "fail if given a class" in {
      intercept[IllegalArgumentException] { 
        val p = Proxy.create(threadContextClassLoaderStrategy, classOf[Class1]) { 
          unimplemented _
        }
      }
    }
    
    "forward calls" in {
      val p = Proxy.create(threadContextClassLoaderStrategy, classOf[Trait1]) { 
        (proxy: AnyRef, name: Symbol, args: Array[AnyRef]) =>
          expect(name) { 'method1 }
          expect(2) { args.length }
          expect(42) { args(0).asInstanceOf[Int] }
          expect("foo") { args(1).asInstanceOf[String] }
          "called"
      }.asInstanceOf[Trait1]

      expect("called") { p.method1(42, "foo") }
    }
  }
}


core_tests/src/test/scala/ProxyMockTest.scala
// Copyright (c) 2011 Paul Butcher
// 
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
// 
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
// 
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.

package com.borachio

import com.borachio.scalatest.MockFactory
import org.scalatest.Suite

class ProxyMockTest extends Suite with MockFactory {
  
  autoVerify = false
  
  trait Turtle {
    def penUp()
    def penDown()
    def forward(distance: Double): (Double, Double)
    def turn(angle: Double)
    def getAngle: Double
    def getPosition(): (Double, Double)
    def setPosition(x: Double, y: Double): (Double, Double)
    def moveInSequence(positions: (Double, Double)*)
  }
  
  def testUnexpectedCall {
    val m = mock[Turtle]
    intercept[ExpectationException] { m.penDown }
  }

  def testSingleExpectation {
    val m = mock[Turtle]
    m expects 'setPosition withArguments (1.0, 2.0) returning (3.0, 4.0)
    expect((3.0, 4.0)) { m.setPosition(1.0, 2.0) }
    verifyExpectations
  }
  
  def testMultipleExpectationsOnMultipleObjects {
    val m1 = mock[Turtle]
    val m2 = mock[Turtle]
    
    inSequence {
      m1 expects 'setPosition withArguments (0.0, 0.0)
      m1 expects 'penDown
      m1 expects 'forward withArguments (10.0)
      m1 expects 'penUp
    }
    inSequence {
      m2 expects 'setPosition withArguments(1.0, 1.0)
      m2 expects 'turn withArguments (90.0)
      m2 expects 'forward withArguments (1.0)
      m2 expects 'getPosition returning (2.0, 1.0)
    }
    
    m2.setPosition(1.0, 1.0)
    m1.setPosition(0.0, 0.0)
    m1.penDown
    m2.turn(90.0)
    m1.forward(10.0)
    m2.forward(1.0)
    m1.penUp
    expect((2.0, 1.0)) { m2.getPosition }

    verifyExpectations
  }
  
  def testRepeatedParameters {
    val m = mock[Turtle]
    
    m expects 'moveInSequence withArguments(Seq((1.0, 1.0)))
    m expects 'moveInSequence withArguments(Seq((2.0, 2.0), (-1.0, -2.0), (10.0, 0.0)))
    
    m.moveInSequence((1.0, 1.0))
    m.moveInSequence((2.0, 2.0), (-1.0, -2.0), (10.0, 0.0))
    
    verifyExpectations
  }
  
  def testStubs {
    val m = mock[Turtle]
    
    m stubs 'setPosition
    m stubs 'getPosition returning (3.0, 4.0)
    m stubs 'forward

    m.setPosition(1.0, 2.0)
    expect((3.0, 4.0)) { m.getPosition }
    m.setPosition(5.0, 6.0)
    
    verifyExpectations
  }
  
  def testPredicate {
    val m = mock[Turtle]
    
    m expects 'setPosition where { (x: Double, y: Double) => x == y } anyNumberOfTimes
    
    m.setPosition(1.0, 1.0)
    m.setPosition(3.14159, 3.14159)
    intercept[ExpectationException] { m.setPosition(1.0, 2.0) }
  }
  
  def testMockReturningMock {
    trait Parent {
      def getChild: Child
    }
    trait Child
    
    val p = mock[Parent]
    val c = mock[Child]
    
    p expects 'getChild returning c
    
    expect(c) { p.getChild }
    
    verifyExpectations
  }
}


core/src/main/scala/ProxyMockFactory.scala
// Copyright (c) 2011 Paul Butcher
// 
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
// 
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
// 
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.

package com.borachio

import scala.collection.mutable.{ListBuffer, Map}

trait ProxyMockFactory { self: AbstractMockFactory =>
  
  protected var proxyClassLoaderStrategy = threadContextClassLoaderStrategy
  
  protected def mock[T: ClassManifest] = {
    val proxy = Proxy.create(proxyClassLoaderStrategy, classOf[Mock], classManifest[T].erasure) {
      (proxy: AnyRef, name: Symbol, args: Array[AnyRef]) =>
        try {
          name match {
            case 'expects => addExpectation(proxy, args(0).asInstanceOf[Symbol])
            case 'stubs => addExpectation(proxy, args(0).asInstanceOf[Symbol]).anyNumberOfTimes
            case 'toString => "proxy mock object "+ System.identityHashCode(proxy)
            case _ => methodsFor(proxy)(name)(args).asInstanceOf[AnyRef]
          }
        } catch {
          case e: NoSuchElementException => {
            val argsString = if (args != null)
              " with arguments: "+ args.mkString("(", ", ", ")")
            else
              ""
            self.handleUnexpectedCall(name + argsString)
          }
        }
    }
    proxies += (proxy -> Map[Symbol, ProxyMockFunction]())
    proxy.asInstanceOf[T with Mock]
  }
  
  private def methodsFor(proxy: AnyRef) = (proxies find { _._1 eq proxy }).get._2
  
  private def addExpectation(proxy: AnyRef, method: Symbol) =
    self.MockFunctionToExpectation(getOrCreate(proxy, method))
  
  private def getOrCreate(proxy: AnyRef, name: Symbol) = {
    val methods = methodsFor(proxy)
    if (methods contains name) {
      methods(name)
    } else {
      val m = new ProxyMockFunction(name, self)
      methods += name -> m
      m
    }
  }

  private val proxies = ListBuffer[(AnyRef, Map[Symbol, ProxyMockFunction])]()
}


project/Build.scala
// Copyright (c) 2011 Paul Butcher
// 
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
// 
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
// 
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.

import sbt._
import Keys._

object Borachio extends Build {

  override lazy val settings = super.settings ++ Seq(
    organization := "com.borachio",
    version := "1.5",
    crossScalaVersions := Seq("2.8.1", "2.9.0", "2.9.0-1", "2.9.1"),

    scalacOptions ++= Seq("-deprecation", "-unchecked", "-Xfatal-warnings"),

    publishTo <<= version { v =>
      val nexus = "http://nexus.scala-tools.org/content/repositories/"
      if (v.trim.endsWith("SNAPSHOT"))
        Some("snapshots" at nexus + "snapshots/") 
      else
        Some("releases" at nexus + "releases/")
    },
    credentials += Credentials(Path.userHome / ".ivy2" / ".credentials")
  )
  
  lazy val core = Project("Borachio core", file("core"), 
    aggregate = Seq(scalatest, specs2, junit3, core_tests)) settings(
    // Workaround https://github.com/harrah/xsbt/issues/193
    unmanagedClasspath in Compile += Attributed.blank(new java.io.File("doesnotexist")))

  lazy val scalatest: Project = Project("Borachio ScalaTest support", file("frameworks/scalatest")) settings(
    libraryDependencies <+= scalaVersion("org.scalatest" %% "scalatest" % scalatestVersion(_))
  ) dependsOn(core)

  lazy val specs2: Project = Project("Borachio Specs2 support", file("frameworks/specs2")) settings(
    libraryDependencies <+= scalaVersion("org.specs2" %% "specs2" % specs2Version(_))
  ) dependsOn(core)

  lazy val junit3: Project = Project("Borachio Junit3 support", file("frameworks/junit3")) settings(
    libraryDependencies += "junit" % "junit" % "3.8.2"
  ) dependsOn(core)
  
  lazy val core_tests: Project = Project("Tests", file("core_tests"), 
    dependencies = Seq(scalatest % "test", specs2 % "test")) settings(publish := ())
  
  val scalatestVersions = Map("2.8" -> "1.5.1", "2.9" -> "1.6.1")
  val specs2Versions = Map("2.8" -> "1.2", "2.9" -> "1.5")
    
  def scalatestVersion(scalaVersion: String) = getLibraryVersion(scalatestVersions, scalaVersion)
  def specs2Version(scalaVersion: String) = getLibraryVersion(specs2Versions, scalaVersion)
  
  def getLibraryVersion(versionMap: Map[String, String], scalaVersion: String) =
    versionMap.getOrElse(scalaVersion take 3, sys.error("Unsupported Scala version: "+ scalaVersion))
}
